sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,276 @@
+with different `module-id` attribute, the second one is not copied to the
+device image with the first function, but dependency between those device images
+is recorder instead.
+
+After `SYCL_EXTERNAL` functions are arranged into a separate device image(s),
+all non-`SYCL_EXTERNAL` functions are internalized to avoid multiple definition
+errors during runtime linking.
+Device images with `SYCL_EXTERNAL` functions will also have a list of names
+of exported functions.
+
+**NOTE**: If device code split is enabled, it seems reasonable to perform
+exports arrangement before device code split procedure.
+
+In orger to collect information about imported symbols `sycl-post-link` looks
+through LLVM IR and for each declared but not defined symbol records its name,
+except the following cases:
+- Declarations with `__` prefix in demangled name are not recorded as imported
+  functions
+  - Declarations with `__spirv_*` prefix should not be recorded as dependencies
+  since they represent SPIR-V operations and will be transformed to SPIR-V
+  instructions during LLVM->SPIR-V translation.
+- Based on some attributes which could be defined later
+  - This is needed to have possibility to call device-specific builtins not
+    starting with `__` by forward-declaring them in DPC++ code
+
+**NOTE**: If device code split is enabled, imports collection is performed after
+split and it is performed on splitted images.
+
+All collected information is attached to a device image via properties
+mechanism.
+
+Each device image is supplied with an array of property sets:
+```
+struct pi_device_binary_struct {
+...
+  // Array of property sets
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+
+```
+Each property set is represent by the following struct:
+```
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+List of imported symbols is represented as a single property set with name
+`ImportedSymbols` recorded in the `Name` field of property set.
+Each property in this set holds name of the particular imported symbol recorded
+in the `Name` field of the property.
+List of exported symbols is represented in the same way, except the
+corresponding set has the name `ExportedSymbols`.
+
+### DPC++ runtime changes
+
+DPC++ RT performs *device images collection* task by grouping all device
```

Comment:
I think it is important to mention when it is supposed to happen. 

For SYCL 1.2.1 I would expect this happen only if regular `parallel_for` overload is used and by regular I mean one which does not accept `kernel` object which was extracted from explicitly created `program` object like it is shown in intel/llvm#3180. I.e. if user manipulates somehow with programs - it is responsible to link all required device code himself; if user just wants to launch the kernel, DPC++ runtime will automatically link required code from shared libraries for the user.

I'm not yet familiar enough with SYCL 2020 (especially with sections related to device programs management) to suggest particular way of handling shared libraries there, but I would prefer following the same logic: if user does something with programs, it is then his responsibility to link required device code.

Reply:
I don't think that we don't need to resolve dependencies automatically when SYCL 2020 kernel bundles are involved. Linking of `kernel_bundle` doesn't seem like the same thing as `program` linking from SYCL 1.2.1. `link` API for kernel bundles does not guarantee that all device images from linked kernel bundles will be merged into one device image. Instead it looks like union of sets of device images and then translation to executable state. But, executable state means that all device images from kernel bundle are ready be executed on device and therefore they don't have dependencies. So, it seems like `link` API for kernel bundles does not require linking all underlying device images into one, but it requires symbols resolution if there are any dependencies between device images.